include/boost/gil/image_processing/morphology.hpp

```diff
@@ -0,0 +1,285 @@
+  copy_pixels(src_view, int_op_view);
+  for (int i = 0; i < iterations; ++i)
+    morph(int_op_view, int_op_view, ker_mat, detail::morphological_operation::erosion);
+}
+
+/// \brief Performs erosion and then dilation on the input image view . This
+/// operation is utilized for removing noise from images. \param src_view -
+/// Source/input image view. \param int_op_view - Output image view. \param
+/// ker_mat - Kernel matrix/structuring element containing 0's and 1's which
+/// will be used for applying the opening operation. \tparam SrcView type of
+/// source image, models gil::ImageViewConcept. \tparam IntOpView type of output
+/// image, models gil::MutableImageViewConcept. \tparam Kernel type of
+/// structuring element.
+template <typename SrcView, typename IntOpView, typename Kernel>
+void opening(SrcView const &src_view, IntOpView const &int_op_view,
+             Kernel const &ker_mat) {
+  erode(src_view, int_op_view, ker_mat, 1);
+  dilate(int_op_view, int_op_view, ker_mat, 1);
+}
+
+/// \brief Performs dilation and then erosion on the input image view which is
+/// exactly opposite to the opening operation . Closing operation can be
+/// utilized for closing small holes inside foreground objects. \param src_view
+/// - Source/input image view. \param int_op_view - Output image view. \param
+/// ker_mat - Kernel matrix/structuring element containing 0's and 1's which
+/// will be used for applying the closing operation. \tparam SrcView type of
+/// source image, models gil::ImageViewConcept. \tparam IntOpView type of output
+/// image, models gil::MutableImageViewConcept. \tparam Kernel type of
+/// structuring element.
+template <typename SrcView, typename IntOpView, typename Kernel>
+void closing(SrcView const &src_view, IntOpView const &int_op_view,
+             Kernel const &ker_mat) {
+  dilate(src_view, int_op_view, ker_mat, 1);
+  erode(int_op_view, int_op_view, ker_mat, 1);
+}
+
+/// \brief Calculates the difference between image views generated after
+/// applying dilation dilation and erosion on an image . The resultant image
+/// will look like the outline of the object(s) present in the image. \param
+/// src_view - Source/input image view. \param dst_view - Destination view which
+/// will store the final result of morphological gradient operation. \param
+/// ker_mat - Kernel matrix/structuring element containing 0's and 1's which
+/// will be used for applying the morphological gradient operation. \tparam
+/// SrcView type of source image, models gil::ImageViewConcept. \tparam DstView
+/// type of output image, models gil::MutableImageViewConcept. \tparam Kernel
+/// type of structuring element.
+template <typename SrcView, typename DstView, typename Kernel>
+void morphological_gradient(SrcView const &src_view, DstView const &dst_view,
+                            Kernel const &ker_mat) {
+  using namespace boost::gil;
+  gray8_image_t int_dilate(src_view.dimensions()),
```

Comment:
Not all views will be gray image views, you can try `gil::image<typename DstView::value_type>`. Same thing with hats.